File: packages/core/src/server/ripgrep.ts
import { execFile } from 'node:child_process';
import path from 'node:path';
import type { SearchContentResult } from '@qlan-ro/mainframe-types';
import { createChildLogger } from '../logger.js';

const logger = createChildLogger('ripgrep');

const MAX_LINE_LENGTH = 500;
const TIMEOUT_MS = 30_000;

export function parseRipgrepOutput(output: string, basePath: string, maxResults = 200): SearchContentResult[] {
  const results: SearchContentResult[] = [];

  for (const line of output.split('\n')) {
    if (results.length >= maxResults) break;
    if (!line.trim()) continue;

    let parsed: { type: string; data?: unknown };
    try {
      parsed = JSON.parse(line) as { type: string; data?: unknown };
    } catch {
      continue;
    }

    if (parsed.type !== 'match' || !parsed.data) continue;

    const data = parsed.data as {
      path?: { text?: string };
      line_number?: number;
      lines?: { text?: string };
      submatches?: Array<{ start?: number }>;
    };

    const filePath = data.path?.text;
    const lineNumber = data.line_number;
    const lineText = data.lines?.text;
    const submatches = data.submatches;

    if (!filePath || !lineNumber || lineText == null) continue;

    const relFile = path.relative(basePath, filePath);
    const text = lineText.replace(/\n$/, '').slice(0, MAX_LINE_LENGTH);
    const column = (submatches?.[0]?.start ?? 0) + 1;

    results.push({ file: relFile, line: lineNumber, column, text });
  }

  return results;
}

let rgBinaryPath: string | null | undefined = undefined;

function getRgPath(): string | null {
  if (rgBinaryPath !== undefined) return rgBinaryPath;
  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const { rgPath } = require('@vscode/ripgrep') as { rgPath: string };
    rgBinaryPath = rgPath;
    return rgBinaryPath;
  } catch {
    logger.warn('Failed to load @vscode/ripgrep — ripgrep search unavailable');
    rgBinaryPath = null;
    return null;
  }
}

export interface RipgrepOptions {
  maxResults?: number;
  maxFileSize?: string;
  includeIgnored?: boolean;
}

export function searchWithRipgrep(
  scopePath: string,
  query: string,
  opts?: RipgrepOptions,
): Promise<SearchContentResult[]> {
  const rgPath = getRgPath();
  if (!rgPath) return Promise.resolve([]);

  const maxResults = opts?.maxResults ?? 200;
  const maxFileSize = opts?.maxFileSize ?? '1M';

  const args = ['--json', '--ignore-case', '--max-filesize', maxFileSize, '--no-require-git', '--max-count', '50'];

  if (opts?.includeIgnored) {
    args.push('--no-ignore', '--hidden');
  }

  args.push('--', query, scopePath);

  return new Promise((resolve) => {
    const child = execFile(
      rgPath,
      args,
      { timeout: TIMEOUT_MS, maxBuffer: 10 * 1024 * 1024 },
      (err, stdout, stderr) => {
        if (stderr) {
          logger.debug({ stderr: stderr.slice(0, 500) }, 'ripgrep stderr');
        }

        // Exit code 1 = no matches (not an error). Exit code 2 = partial error.
        // execFile wraps process exit codes in err with numeric status on err.
        type ExecErr = NodeJS.ErrnoException & { killed?: boolean };
        const execErr = err as ExecErr | null;
        const isNoMatches = execErr && String((execErr as { code?: unknown }).code) === '1';
        if (execErr && !isNoMatches) {
          logger.warn({ err }, 'ripgrep process error');
        }

        if (!stdout) {
          resolve([]);
          return;
        }

        resolve(parseRipgrepOutput(stdout, scopePath, maxResults));
      },
    );

    // Safety timeout — kill if still running after TIMEOUT_MS
    setTimeout(() => {
      if (!child.killed) child.kill('SIGTERM');
    }, TIMEOUT_MS);
  });
}

export function listFilesWithRipgrep(dirPath: string, opts?: { includeIgnored?: boolean }): Promise<string[] | null> {
  const rgPath = getRgPath();
  if (!rgPath) return Promise.resolve(null);

  const args = ['--files', '--no-require-git'];

  if (opts?.includeIgnored) {
    args.push('--no-ignore', '--hidden');
  }

  args.push(dirPath);

  return new Promise((resolve) => {
    execFile(rgPath, args, { timeout: TIMEOUT_MS, maxBuffer: 10 * 1024 * 1024 }, (err, stdout, stderr) => {
      if (stderr) {
        logger.debug({ stderr: stderr.slice(0, 500) }, 'ripgrep --files stderr');
      }

      if (err) {
        logger.warn({ err }, 'ripgrep --files failed');
        resolve(null);
        return;
      }

      const files = stdout
        .split('\n')
        .map((line) => line.trim())
        .filter((line) => line.length > 0)
        .map((absPath) => path.relative(dirPath, absPath));

      resolve(files);
    });
  });
}

export function isRipgrepAvailable(): boolean {
  return getRgPath() !== null;
}

File: packages/core-rs/crates/mainframe-server/src/ripgrep.rs
//! Ported from `src/server/ripgrep.ts`.
//!
//! Content and file-name search reimplemented in-process on ripgrep's own
//! library crates (`ignore` + `grep-searcher` + `grep-regex`, pinned to the
//! versions ripgrep 14.1.1 itself vendors) instead of shelling out to a
//! resolved `rg` binary. Search now always works — there is no more "ripgrep
//! unavailable, fall back to a JS walk" branch anywhere in the routes.

use std::path::Path;

use grep_regex::RegexMatcherBuilder;
use grep_searcher::{BinaryDetection, SearcherBuilder};
use ignore::WalkBuilder;
use mainframe_types::search::SearchContentResult;

use crate::fs_utils::{is_ignored_dir, relative};
use sink::ContentSink;

mod sink;

const DEFAULT_MAX_RESULTS: usize = 200;
const DEFAULT_MAX_FILE_SIZE: u64 = 1024 * 1024; // "1M", matches the old CLI default.

/// Options for [`search_with_ripgrep`]. `None` fields take the defaults above.
#[derive(Debug, Clone, Default)]
pub struct RipgrepOptions {
    pub max_results: Option<usize>,
    pub max_file_size: Option<String>,
    pub include_ignored: bool,
}

/// Case-insensitive content search under `scope_path`, capped at
/// `max_results` total hits (50 per file). Gitignore/hidden-file rules apply
/// unless `include_ignored`. Runs off the async runtime via `spawn_blocking`
/// since `ignore`/`grep-searcher` are synchronous.
pub async fn search_with_ripgrep(
    scope_path: &str,
    query: &str,
    opts: &RipgrepOptions,
) -> Vec<SearchContentResult> {
    let scope_path = scope_path.to_string();
    let query = query.to_string();
    let opts = opts.clone();
    tokio::task::spawn_blocking(move || search_blocking(&scope_path, &query, &opts))
        .await
        .unwrap_or_default()
}

fn search_blocking(
    scope_path: &str,
    query: &str,
    opts: &RipgrepOptions,
) -> Vec<SearchContentResult> {
    let matcher = match RegexMatcherBuilder::new()
        .case_insensitive(true)
        .build(query)
    {
        Ok(m) => m,
        Err(err) => {
            tracing::warn!(error = %err, query, "invalid search pattern");
            return Vec::new();
        }
    };
    let max_results = opts.max_results.unwrap_or(DEFAULT_MAX_RESULTS);
    let max_file_size = opts
        .max_file_size
        .as_deref()
        .map(parse_max_file_size)
        .unwrap_or(Some(DEFAULT_MAX_FILE_SIZE));

    let mut builder = WalkBuilder::new(scope_path);
    builder.require_git(false).max_filesize(max_file_size);
    if opts.include_ignored {
        disable_ignore_rules(&mut builder);
    }

    let mut searcher = SearcherBuilder::new()
        .line_number(true)
        .binary_detection(BinaryDetection::quit(0))
        .build();
    let base = Path::new(scope_path);
    let mut results = Vec::new();

    for entry in builder.build() {
        if results.len() >= max_results {
            break;
        }
        let Ok(entry) = entry else { continue };
        if !entry.file_type().map(|ft| ft.is_file()).unwrap_or(false) {
            continue;
        }
        let rel_file = relative(base, entry.path());
        let mut sink = ContentSink {
            matcher: &matcher,
            rel_file: &rel_file,
            results: &mut results,
            max_results,
            hits_in_file: 0,
        };
        if let Err(err) = searcher.search_path(&matcher, entry.path(), &mut sink) {
            tracing::warn!(error = %err, file = %entry.path().display(), "content search failed for file");
        }
    }
    results
}

/// Disables every ignore mechanism (`.gitignore`, global/local excludes,
/// hidden-file skip) — mirrors the old CLI's `--no-ignore --hidden`.
fn disable_ignore_rules(builder: &mut WalkBuilder) {
    builder
        .hidden(false)
        .ignore(false)
        .git_ignore(false)
        .git_global(false)
        .git_exclude(false)
        .parents(false);
}

/// Parses the CLI-style `max_file_size` string (`"1M"`, `"512K"`, or a bare
/// byte count) the old code passed straight to `rg --max-filesize`. `None` on
/// empty/unparseable input (the walker then applies no size cap).
fn parse_max_file_size(raw: &str) -> Option<u64> {
    let raw = raw.trim();
    let (digits, multiplier) = match raw.chars().last() {
        Some(c) if c.eq_ignore_ascii_case(&'k') => (&raw[..raw.len() - 1], 1024),
        Some(c) if c.eq_ignore_ascii_case(&'m') => (&raw[..raw.len() - 1], 1024 * 1024),
        Some(c) if c.eq_ignore_ascii_case(&'g') => (&raw[..raw.len() - 1], 1024 * 1024 * 1024),
        _ => (raw, 1),
    };
    digits.trim().parse::<u64>().ok().map(|n| n * multiplier)
}

/// Options for [`list_files_with_ripgrep`].
#[derive(Debug, Clone, Default)]
pub struct ListFilesOptions {
    pub include_ignored: bool,
    pub use_builtin_ignore_only: bool,
}

/// Lists files under `dir_path`, relative to it. `use_builtin_ignore_only`
/// skips `.gitignore` (surfacing gitignored config files like `.env`) but
/// still excludes the project's ignored directories; `include_ignored`
/// disables every ignore rule outright.
pub async fn list_files_with_ripgrep(dir_path: &str, opts: &ListFilesOptions) -> Vec<String> {
    let dir_path = dir_path.to_string();
    let opts = opts.clone();
    tokio::task::spawn_blocking(move || list_files_blocking(&dir_path, &opts))
        .await
        .unwrap_or_default()
}

fn list_files_blocking(dir_path: &str, opts: &ListFilesOptions) -> Vec<String> {
    let mut builder = WalkBuilder::new(dir_path);
    builder.require_git(false);
    if opts.use_builtin_ignore_only {
        disable_ignore_rules(&mut builder);
        // .gitignore is off, so surface gitignored config files — but still
        // skip build-artifact dirs (node_modules, target, ...) by name.
        builder.filter_entry(|entry| {
            entry.depth() == 0
                || entry
                    .file_name()
                    .to_str()
                    .map(|name| !is_ignored_dir(name))
                    .unwrap_or(true)
        });
    } else if opts.include_ignored {
        disable_ignore_rules(&mut builder);
    }

    let base = Path::new(dir_path);
    builder
        .build()
        .filter_map(Result::ok)
        .filter(|e| e.file_type().map(|ft| ft.is_file()).unwrap_or(false))
        .map(|e| relative(base, e.path()))
        .collect()
}

#[cfg(test)]
mod tests;

// PORT STATUS: src/server/ripgrep.ts (parseRipgrepOutput, getRgPath,
// searchWithRipgrep, listFilesWithRipgrep, isRipgrepAvailable)
// confidence: high
// todos: 0
// notes: Rewritten for PR 1 of the Rust-daemon cutover onto ripgrep's own
// library crates (`ignore` + `grep-searcher` + `grep-regex`, pinned to the
// exact versions ripgrep 14.1.1 vendors) instead of shelling out to a resolved
// `rg` binary. Search always runs in-process now, so `isRipgrepAvailable`,
// binary resolution, and `MAINFRAME_RG_PATH` have no successor — every
// caller's "unavailable, fall back to a walk" branch became unreachable and
// was deleted along with it (routes/search.rs's `search_directory_fallback`,
// routes/files.rs's `search_walk`). `--no-require-git` -> `WalkBuilder::
// require_git(false)`; `--no-ignore --hidden` -> `disable_ignore_rules`;
// `--max-count 50` -> the sink's per-file `hits_in_file` cap; `--max-filesize`
// -> `parse_max_file_size` + `WalkBuilder::max_filesize`. Both entry points
// move the synchronous walk/search onto `spawn_blocking` (the daemon forbids
// sync I/O on the async runtime). `line.slice(0,500)` (UTF-16 units) -> first
// 500 chars (UTF-8 safe), same as before.
